core/arch/x86/mangle.c

```diff
@@ -535,10 +594,48 @@ insert_pop_all_registers(dcontext_t *dcontext, clean_call_info_t *cci, instrlist
                                               OPSZ_SAVED_XMM)));
             }
         }
-        ASSERT(i * MCXT_SIMD_SLOT_SIZE == MCXT_TOTAL_SIMD_SLOTS_SIZE ||
-               i * MCXT_SIMD_SLOT_SIZE == MCXT_TOTAL_SIMD_SSE_AVX_SLOTS_SIZE);
+        if (ZMM_ENABLED()) {
+            PRE(ilist, instr, INSTR_CREATE_jmp(dcontext, opnd_create_instr(post_pop)));
+            PRE(ilist, instr, pre_avx512_pop /*label*/);
+            uint opcode_avx512 = move_mm_avx512_reg_opcode(ALIGNED(alignment, 64));
+            for (i = 0; i < proc_num_simd_registers(); i++) {
+                if (!cci->simd_skip[i]) {
+                    instr_t *simdmov = instr_create_1dst_2src(
+                        dcontext, opcode_avx512,
+                        opnd_create_reg(DR_REG_START_ZMM + (reg_id_t)i),
+                        opnd_create_reg(DR_REG_K0),
+                        opnd_create_base_disp(REG_XSP, REG_NULL, 0,
+                                              PRE_XMM_PADDING + i * MCXT_SIMD_SLOT_SIZE +
+                                                  offs_beyond_xmm,
+                                              OPSZ_SAVED_ZMM));
+                    PRE(ilist, instr, simdmov);
+                }
+            }
+            for (i = 0; i < proc_num_opmask_registers(); i++) {
+                if (!cci->opmask_skip[i]) {
+                    instr_t *maskmov = instr_create_1dst_1src(
+                        dcontext,
+                        proc_has_feature(FEATURE_AVX512BW) ? OP_kmovq : OP_kmovw,
+                        opnd_create_reg(DR_REG_START_OPMASK + (reg_id_t)i),
+                        opnd_create_base_disp(REG_XSP, REG_NULL, 0,
+                                              PRE_XMM_PADDING +
+                                                  MCXT_TOTAL_SIMD_SLOTS_SIZE +
+                                                  offs_beyond_xmm,
+                                              OPSZ_SAVED_OPMASK));
+                    PRE(ilist, instr, maskmov);
+                }
+            }
+            PRE(ilist, instr, post_pop /*label*/);
+        }
+    }
+
+    if (!cci->skip_save_flags) {
+        PRE(ilist, instr, INSTR_CREATE_popf(dcontext));
+        offs_beyond_xmm = XSP_SZ; /* pc */
+        ;
+    } else if (cci->preserve_mcontext) {
+        offs_beyond_xmm = 2 * XSP_SZ; /* aflags + pc */
```

Comment:
It's already set to this.  A little confusing like this.  I would say remove this whole else.